### PR TITLE
Add an IMU sensor plugin that inherits from SensorPlugin

### PR DESCRIPTION
**Pull request by alessandrosettimi**
We added a new imu plugin: gazebo_ros_imu_sensor.
Main differences from gazebo_ros_imu:
- inheritance from SensorPlugin,
- measurements are given from gazebo imu sensor instead of being computed by the ros plugin,
- gravity is included in inertial measurements.

### CODE REVIEW

**Comment by l0g1x** [comment]
Has anyone had time to look at this?

**Comment by hsu** [comment]
This looks good. Could you please add your overview (3 bullet points in PR description) on differences between gazebo_ros_imu and gazebo_ros_imu_sensor somewhere? I don't have a good place for it in mind (maybe [here](http://gazebosim.org/tutorials?tut=ros_gzplugins#IMU)). I am looking into doxygen generation as well.

**Comment by alessandrosettimi** [comment]
I uptaded the documentation of the plugin and I also created this pull request in the tutorials documentation: https://bitbucket.org/osrf/gazebo_tutorials/pull-requests/335

**Comment by alessandrosettimi** [comment]
The aforementioned pull request has been merged.

**Comment by nicolaerosia** [comment]
This is present in tutorial but not merged: http://gazebosim.org/tutorials?tut=ros_gzplugins&cat=connect_ros#IMUsensor(GazeboRosImuSensor)

**Comment by j-rivero** [comment]
@osrf-jenkins run test please

File: gazebo_plugins/src/gazebo_ros_imu_sensor.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  accelerometer_data = math::Vector3(0, 0, 0);
+  gyroscope_data = math::Vector3(0, 0, 0);
+  orientation = math::Quaternion(1,0,0,0);
+  seed=0;
+}
+
+void gazebo::GazeboRosImuSensor::Load(gazebo::sensors::SensorPtr sensor_, sdf::ElementPtr sdf_)
+{
+  sdf=sdf_;
+  sensor=dynamic_cast<gazebo::sensors::ImuSensor*>(sensor_.get());
```
Comment (meyerj) [inline gazebo_plugins/src/gazebo_ros_imu_sensor.cpp:36]: You should check against null pointer after the dynamic_cast here. Maybe the plugin was loaded for a sensor which is not an `ImuSensor`.

**Comment by meyerj** [comment]
Is there a reason why you took over the scalar `gaussianNoise` parameter from `gazebo_ros_imu` and add a white noise term to all fields individually instead of relying on the more powerful noise model already present in Gazebo's `ImuSensor`? Using the same noise variance for linear acceleration, angular velocity and even quaternion components has nothing do to with reality. However, the current `ImuSensor` (https://bitbucket.org/osrf/gazebo/src/2e153b94b73a5f3e6301cc65640920cdb311f08c/gazebo/sensors/ImuSensor.cc?at=default&fileviewer=file-view-default) does not apply any noise or bias to the orientation part, which is also not very realistic.

+1 for providing an alternative ROS plugin for the broken `gazebo_ros_imu` plugin (unrealistic noise and no gravity component in acceleration).

**Comment by 130s** [comment]
https://github.com/ros-simulation/gazebo_ros_pkgs/pull/363#issuecomment-204942236

> This is present in tutorial but not merged: http://gazebosim.org/tutorials?tut=ros_gzplugins&cat=connect_ros#IMUsensor(GazeboRosImuSensor)

Yeah. Just as fyi, the tutorial is confusing (2 IMU plugins. One is not existent yet (`GazeboRosImuSensor`. The one this PR suggests). The info for the other `GazeboRosImu` isn't really helpful). Since we still have only `GazeboRosImu`, I [opened a pr to update the doc](https://bitbucket.org/osrf/gazebo_tutorials/pull-requests/370/update-for-imu-plugin-tutorial/diff).

File: gazebo_plugins/src/gazebo_ros_imu_sensor.cpp (hunk 1 above)
Comment (alessandrosettimi) [inline gazebo_plugins/src/gazebo_ros_imu_sensor.cpp:36]: I added the test you suggested.

**Comment by alessandrosettimi** [comment]
I added a check on Gazebo version since some functions have changed since the version 6.0

**Comment by alessandrosettimi** [comment]
I also changed some function that are now deprecated for gazebo version greater than 7.0.

**Comment by jbrookshire** [comment]
Yes, my thanks too for adding this sensor plugin.  I agree, though: not clear why `gaussianNoise` parameter  is needed.  It seems to just add Gaussian noise on top of the sensor's Gaussian noise.

File: gazebo_plugins/src/gazebo_ros_imu_sensor.cpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+bool gazebo::GazeboRosImuSensor::LoadParameters()
+{
+  //loading parameters from the sdf file
+
+  //NAMESPACE
+  if (sdf->HasElement("robotNamespace"))
+  {
+    robot_namespace =  "/" + sdf->Get<std::string>("robotNamespace") +"/";
```
Comment (sarlinpe) [inline gazebo_plugins/src/gazebo_ros_imu_sensor.cpp]: Why not including the root '/' in the namespace ? As already used in the gazebo_ros_control plugin (see http://gazebosim.org/tutorials?tut=ros_control&cat=connect_ros)

File: gazebo_plugins/src/gazebo_ros_imu_sensor.cpp [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    ROS_INFO_STREAM("<topicName> set to: "<<topic_name);
+  }
+  else
+  {
+    topic_name = robot_namespace + "/imu_data";
+    ROS_WARN_STREAM("missing <topicName>, set to /namespace/default: " << topic_name);
+  }
+
+  //BODY NAME
+  if (sdf->HasElement("bodyName"))
```
Comment (sarlinpe) [inline gazebo_plugins/src/gazebo_ros_imu_sensor.cpp]: "bodyName" is referenced as "frameName" in other plugins (see [here](http://gazebosim.org/tutorials?tut=ros_gzplugins&cat=connect_ros#Camera)), maybe better keeping the same convention ?

**Review by davetcoleman — APPROVED** [review]

File: gazebo_plugins/src/gazebo_ros_imu_sensor.cpp (hunk 2 above)
Comment (alessandrosettimi) [reply gazebo_plugins/src/gazebo_ros_imu_sensor.cpp]: done in 3b4735a

File: gazebo_plugins/src/gazebo_ros_imu_sensor.cpp (hunk 3 above)
Comment (alessandrosettimi) [reply gazebo_plugins/src/gazebo_ros_imu_sensor.cpp]: done in 0cb04ca

**Comment by davetcoleman** [comment]
@j-rivero think this can be merged? I'm not super familiar with IMU sensors. This PR also targets indigo, maybe it should just be targeted to Kinetic since the original PR is so old now?

**Comment by iluetkeb** [comment]
+1 for merging. This can only be better than what's currently available.

Lets merge first to Indigo and then pull up to Kinetic.

**Comment by jliviero** [comment]
We would also find this plugin useful, +1 to merging.

**Comment by j-rivero** [comment]
Thanks for the feedback. Before merging, it would be nice to include a test or at least an example world that includes the IMU so black box testing can be run without having to implement one.

@iluetkeb we try to merge and release new additions or modifications first into Jade/Kinetic and threat Indigo as a more stable code where changes landed after a period of testing in the other distributions.

**Comment by j-rivero** [comment]
Thanks very much @alessandrosettimi for the work and the rest of you for the comments.

**Pull request merged**